### DISCUSSION TITLE
What do all the Failure Codes mean?

### DISCUSSION THREAD

**Discussion by ItsSour** [PartialVolume/shredos.x86_64#90] [Q&A] [answered]
What are all the failure codes that are programed into the OS and what do they all mean? This would be good to know in case I have to do some troubleshooting.

**Comment by PartialVolume** [comment]
I assume you mean errors that the nwipe code generates?

It doesn't tend to produce distinct exit codes, generally -1 for error, 0 for success. All errors have a fairly obvious & verbose English error message written to the nwipe log which is located in the root directory. So the first place to look is in that log file. The log file is only created on disc on nwipe exit, however the tail end text & summary tables are automatically shown on screen on exit. To view the full log you would jump to the second virtual terminal ALT F2 and type ls to list the files. Choose the most recent nwipe[date&time].log and use more/cat to display it.

Other than a I/O errors related to failing discs, I don't tend to see errors in general usage as the code is pretty stable.

If anybody wants to go through the code and compile a list of all possible error messages and post it here then please feel free.

**Reply by ItsSour** [reply]
I'll touch up on my linux and thumb through it, thanks!

**Comment by ItsSour** [comment]
So I tried to get some of them, I don't think I got all of them though. I will attach a document later with everything compiled.

As far as I can tell for the GUI, the only error displayed in the GUI is "Failure Code -1" which actually ties to a linux command `fdatasync`. 'fdatasync' in the command nwipe uses to synch a files in-core state (I imagine this would be determined by the wipe method) with a storage device (anything that holds any bits). It does this by only transferring meta data  needed to allow a subsequent data retrieval to be correctly handled. TL;DR - synchs what the software is trying to do with the hard drive so wiping can continue.

On success, the system calls the return 0, in which nwipe then displays. On error, -1 is returned, and errno (number of last error) is set appropriately. It does have specific errors it can display, but I can't seem to find it anywhere in the code so I don't think its an addition

Either way, nwipe determines what the cause of the error is and writes in the log what it was, but I can't seem to find out how it does that, I looked through a lot of the code for both nwipe and ShredOS and can't find much, i might be looking in the wrong places. Any tips?

Also, are the logs written to the USB thumb drive (if inserted) or are the stuck in the RAM until moved?

Thanks for all the help, I'm still kinda learning about linux processes so if I'm wrong about something just tell me :)

**Comment by PartialVolume** [comment]
>On success, the system calls the return 0, in which nwipe then displays. On error, -1 is returned, and errno (number of last error) is set appropriately. It does have specific errors it can display, but I can't seem to find it anywhere in the code so I don't think its an addition

If I remember correctly, the wipe threads returned 0 on success, -1 on a critical failure that meant the wipe would stop and +1 on a soft error. Trouble is, it was never really clear what a soft error might be. So over time it's become that any error is critical. At least, without looking at the code that how I remember it. I'm not keen on the message Failure code -1, in my opinion it would be better if the threads returned more detailed codes so the appropriate error could be displayed. Something to improve on in the future, however having said that that once the wipe thread is active the device has already been opened so failures that occur will generally be I/O failures picked up by fdatasync.

>Either way, nwipe determines what the cause of the error is and writes in the log what it was, but I can't seem to find out how it does that, I looked through a lot of the code for both nwipe and ShredOS and can't find much, i might be looking in the wrong places. Any tips?

To make a log entry the nwipe_log() function is called always from within the nwipe source code. All logging related functions are located in the source file /src/logging.c and it's header logging.h The logging function works by storing log entries in RAM which are indexed. Mallocs are used to allocate memory for each log message so logging is limited only by memory limits not some arbitrary array size.  Nwipe then calls, on completion of all wipes another logging function to send all log entries to stdout. The allocated memory is then free'ed before nwipe exits. That's how nwipe manages to print log entries that may have happened many hours ago. Nwipe also has a command line option to write the logs to a file, which ShredOS uses. Each wipe is stored in a timestamps nwipe_xxxxx.log file in the root '/' of the RAM drive. Currently ShredOS doesn't copy these logs back to the USB stick however that is a feature that will be added at some point. ShredOS does support ftp'ing those log entries to a FTP server on your network. Instructions in the ShredOS README.md file.

>Thanks for all the help, I'm still kinda learning about linux processes so if I'm wrong about something just tell me :)

I've been using Linux for twenty years and before that UNIX and I'm still learning 😁

**Reply by ItsSour** [reply]
> in my opinion it would be better if the threads returned more detailed codes so the appropriate error could be displayed.

Would that be an issue with the ShredOS code or linux? I can see how it would be a difficult project, maybe a "bandaid" fix would be adding a mode where you can disable the failure code from having an effect of the wipe, so even if there is a critical failure, it will continue attempting to wipe but still log the error. I imagine there would be a pop-up or something warning the user that this could severely damage their hard drive. That way if a hard drive is damaged, you can still remove any sensitive data could be left.

Another thought would be implementing the smartmontools in another way that you guys already do, where if the a critical failure is detected, the system can run a short test (I think its `smartctl -t short dev/sda` or something like that) on the hard drive to give a more detailed error, and depending on the result, ask the user if they wish to continue or not. I'm really just throwing ideas at the wall and seeing if something sticks 😁

> To make a log entry the nwipe_log() function is called always from within the nwipe source code. All logging related functions are located in the source file /src/logging.c and it's header logging.h

I will take a look in those and see what errors can be logged, thanks for the help!

> The logging function works by storing log entries in RAM which are indexed. Mallocs are used to allocate memory for each log message so logging is limited only by memory limits not some arbitrary array size. Nwipe then calls, on completion of all wipes another logging function to send all log entries to stdout. The allocated memory is then free'ed before nwipe exits. That's how nwipe manages to print log entries that may have happened many hours ago. Nwipe also has a command line option to write the logs to a file, which ShredOS uses. Each wipe is stored in a timestamps nwipe_xxxxx.log file in the root '/' of the RAM drive. Currently ShredOS doesn't copy these logs back to the USB stick however that is a feature that will be added at some point. ShredOS does support ftp'ing those log entries to a FTP server on your network. Instructions in the ShredOS README.md file.

This is actually really cool to learn about, thanks for the information, I love how much you guys incorporate the RAM into this project, I haven't seen anything like this being done many other places and its really smart. Keep up the great work!

**Comment by ItsSour** [comment, marked as answer]
@PartialVolume 

So I believe I found an answer to my original question, which was what there error codes mean. This is what nwipe generates but I imagine it's a pretty straight forward crossover to ShredOS. Let me know if this doesn't sound right.

I got a total of 4 messages that can print depending on what happens;
-FAILED- if it failed
 Erased  if it wiped successfully
UABORTED if it was aborted 
INSANITY if there was a bug

The only other thing I found is the use of exclamation tags if there was an error of any kind, or at least that's what I can interpret from this piece of code for the error summary:
`for( i = 0; i < nwipe_selected; i++ )
    {
        if( c[i]->pass_errors != 0 || c[i]->verify_errors != 0 || c[i]->fsyncdata_errors != 0 )
 {
            strncpy( exclamation_flag, "!", 1 );
            exclamation_flag[1] = 0;
        }
        else
        {
            strncpy( exclamation_flag, " ", 1 );
            exclamation_flag[1] = 0;
        }`

Anyways, at this point I'm doing more digging just for fun, I'm pretty sure I found the answer to my original question. Thanks for the help!

**Reply by ItsSour** [reply]
One thing I did notice is that Shred OS failed a wipe, and the nwipe log gave an error log but when i looked at the error summary, there were no errors on any of the drives and it said that the drive had been erased. is this a bug or do they not always match?

**Comment by PartialVolume** [comment]
>One thing I did notice is that Shred OS failed a wipe, and the nwipe log gave an error log but when i looked at the error summary, there were no errors on any of the drives and it said that the drive had been erased. is this a bug or do they not always match?

I'm sure I recently fixed that bug in nwipe [https://github.com/martijnvanbrummelen/nwipe/commit/978c4e9c8bdb31bccd74b6c626ea035ac475ef9e](https://github.com/martijnvanbrummelen/nwipe/commit/978c4e9c8bdb31bccd74b6c626ea035ac475ef9e) and pull request [here](https://github.com/martijnvanbrummelen/nwipe/pull/400)

That fix should be in nwipe & ShredOS 0.32.22 and later. It was only recently committed 12/Jan/2022.

What version of nwipe have you got?

**Reply by ItsSour** [reply]
I'll have to check when I get back to work Monday., but it was the most recent one as of Jan 21st, as that's when I made this account and started using ShredOS.

**Reply by PartialVolume** [reply]
The latest revision was released 28th Jan, so that would probably explain it. You should find it's all fixed in the latest release v2021.08.2_21_x86-64_0.32.023.

**Comment by PartialVolume** [comment]
>The only other thing I found is the use of exclamation tags if there was an error of any kind, or at least that's what I can interpret from this piece of code for the error summary:
for( i = 0; i < nwipe_selected; i++ ) { if( c[i]->pass_errors != 0 || c[i]->verify_errors != 0 || c[i]->fsyncdata_errors != 0 ) { strncpy( exclamation_flag, "!", 1 ); exclamation_flag[1] = 0; } else { strncpy( exclamation_flag, " ", 1 ); exclamation_flag[1] = 0; }

That's correct, In the summary table and error log, as well as a text entry stating success or failure a exclamation mark is placed as the first character of the line to highlight that the line contains an error.

```
/* Example */

!     sdb |--FAIL--|  120MB/s | 01:22:01 | WD6788.8488YNHj/ZX677888388-N
      sdc | Erased |  120MB/s | 01:25:04 | WD6784.8488JKGG/ZX677888388-N
```
